File: CS/MultiRangeTrackBarControlExample/MultiRangeTrackBar/MultiTrackBarViewInfo.cs
using DevExpress.Utils.Drawing;
using DevExpress.XtraEditors.Drawing;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.ViewInfo;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;

namespace MultiRangeTrackBarControlExample {
    public class MultiTrackBarViewInfo : TrackBarViewInfo {
        public MultiTrackBarViewInfo(RepositoryItem item) : base(item) {
            CreateThumbs();
        }

        public override TrackBarObjectPainter GetTrackPainter() {
            return new MultiTrackBarObjectPainter(LookAndFeel);
        }

        public new RepositoryItemMultiTrackBar Item => base.Item as RepositoryItemMultiTrackBar;

        public CustomThumb[] Thumbs { get; protected set; }

        public override void CalcViewInfo(Graphics g) {
            base.CalcViewInfo(g);
            CreateThumbs();
        }

        public override void Offset(int x, int y) {
            base.Offset(x, y);
            foreach (CustomThumb thumb in Thumbs) {
                Point p = thumb.Position;
                p.Offset(x, y);
                thumb.Position = p;

                Rectangle b = thumb.Bounds;
                b.Offset(x, y);
                thumb.Bounds = b;
            }
        }

        public void CreateThumbs(object editValue = null) {
            var list = EditValue as IEnumerable<int>;
            if (list == null) {
                Thumbs = new CustomThumb[0];
                return;
            }

            Thumbs = list.Select(CalcThumb).ToArray();
        }

        public void SetThumbValue(int index, int newValue) {
            Thumbs[index] = CalcThumb(newValue);
        }

        public override EditHitInfo CalcHitInfo(Point p) {
            return CreateHitInfo(p);
        }

        protected override ObjectState CalcObjectState() {
            var edit = OwnerEdit as MultiRangeTrackBar;
            if (edit == null)
                return ObjectState.Normal;
            if (edit.ActiveThumbIndex != -1)
                return ObjectState.Hot;
            return ObjectState.Normal;
        }

        public void SetThumbPos(Point p) => ThumbPos = p;

        public new MultiTrackBarInfoCalculator TrackCalculator => base.TrackCalculator as MultiTrackBarInfoCalculator;

        private CustomThumb CalcThumb(int value) {
            Point thumbPos = CalcThumbPosCore(value);
            return new CustomThumb {
                Value = value,
                Position = thumbPos,
                Bounds = TrackCalculator.GetThumbBounds(thumbPos)
            };
        }
    }

    public class CustomThumb {
        public int Value { get; set; }
        public Point Position { get; set; }
        public Rectangle Bounds { get; set; }
    }
}

File: VB/MultiRangeTrackBarControlExample/MultiRangeTrackBar/MultiTrackBarViewInfo.vb
Imports DevExpress.Utils.Drawing
Imports DevExpress.XtraEditors.Drawing
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.ViewInfo
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Linq

Namespace MultiRangeTrackBarControlExample
    Public Class MultiTrackBarViewInfo
        Inherits TrackBarViewInfo

        Private _Thumbs As MultiRangeTrackBarControlExample.CustomThumb()

        Public Sub New(ByVal item As RepositoryItem)
            MyBase.New(item)
            CreateThumbs()
        End Sub

        Public Overrides Function GetTrackPainter() As TrackBarObjectPainter
            Return New MultiTrackBarObjectPainter(LookAndFeel)
        End Function

        Public Overloads ReadOnly Property Item As RepositoryItemMultiTrackBar
            Get
                Return TryCast(MyBase.Item, RepositoryItemMultiTrackBar)
            End Get
        End Property

        Public Property Thumbs As CustomThumb()
            Get
                Return _Thumbs
            End Get
            Protected Set(ByVal value As CustomThumb())
                _Thumbs = value
            End Set
        End Property

        Public Overrides Sub CalcViewInfo(ByVal g As Graphics)
            MyBase.CalcViewInfo(g)
            CreateThumbs()
        End Sub

        Public Overrides Sub Offset(ByVal x As Integer, ByVal y As Integer)
            MyBase.Offset(x, y)

            For Each thumb In Thumbs
                Dim p = thumb.Position
                p.Offset(x, y)
                thumb.Position = p
                Dim b = thumb.Bounds
                b.Offset(x, y)
                thumb.Bounds = b
            Next
        End Sub

        Public Sub CreateThumbs()
            Dim list = TryCast(EditValue, IEnumerable(Of Integer))

            If list Is Nothing Then
                Thumbs = New CustomThumb(-1) {}
                Return
            End If

            Thumbs = list.Select(New Func(Of Integer, CustomThumb)(AddressOf CalcThumb)).ToArray()
        End Sub

        Public Sub SetThumbValue(ByVal index As Integer, ByVal newValue As Integer)
            Thumbs(index) = CalcThumb(newValue)
        End Sub

        Public Overrides Function CalcHitInfo(ByVal p As Point) As EditHitInfo
            Return CreateHitInfo(p)
        End Function

        Protected Overrides Function CalcObjectState() As ObjectState
            Dim edit = TryCast(OwnerEdit, MultiRangeTrackBar)
            If edit Is Nothing Then Return ObjectState.Normal
            If edit.ActiveThumbIndex <> -1 Then Return ObjectState.Hot
            Return ObjectState.Normal
        End Function

        Public Sub SetThumbPos(ByVal p As Point)
            ThumbPos = p
        End Sub

        Public Overloads ReadOnly Property TrackCalculator As MultiTrackBarInfoCalculator
            Get
                Return TryCast(MyBase.TrackCalculator, MultiTrackBarInfoCalculator)
            End Get
        End Property

        Private Function CalcThumb(ByVal value As Integer) As CustomThumb
            Dim thumbPos As Point = CalcThumbPosCore(value)
            Return New CustomThumb With {
                .Value = value,
                .Position = thumbPos,
                .Bounds = TrackCalculator.GetThumbBounds(thumbPos)
            }
        End Function
    End Class

    Public Class CustomThumb
        Public Property Value As Integer
        Public Property Position As Point
        Public Property Bounds As Rectangle
    End Class
End Namespace
